### PR TITLE
[PyTorch][Distributed] Use auto-grad enabled collections for the shared linear op to enable backward grad calculation

### LINKED ISSUE #66446: Enable backward pass comms for sharded linear.
## 🚀 Feature
Basic sharded linear implementation was added in https://github.com/pytorch/pytorch/pull/64128, however the backward pass for the comms have not been enabled yet.

## Motivation
For sharded computations for an actual model, forward + backward + optimizer should work for ShardedTensor based ops.

## Pitch
Enabled backward pass for ShardedLinear using https://pytorch.org/docs/stable/distributed.html#autograd-enabled-communication-primitives.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68096
* #68786

We replace all c10d APIs with the Auto-grad collection in the shareded linear op. So that we can enable the backward propagation (grad calculation for sharded linear).

Differential Revision: [D32177341](https://our.internmc.facebook.com/intern/diff/D32177341/)

### CODE REVIEW

File: test/distributed/_sharded_tensor/ops/test_linear.py [hunk 1]
```diff
@@ -63,8 +63,38 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
...
+
+        # Run backward pass
+        local_loss.backward(gradient=external_grad)
+        sharded_loss.backward(gradient=external_grad)
+
+        # Verify that both weight and bias in the sharded linear has non-None grad.
+        sharded_weight = sharded_linear.weight.local_shards()[0].tensor
+        self.assertEqual(sharded_weight.requires_grad, True)
+        self.assertNotEqual(sharded_linear.bias.grad, None)
+        self.assertNotEqual(sharded_weight.grad, None)
```
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/ops/test_linear.py:86]: I think one potential problem might be with the `shard_parameter` method where it doesn't preserve the `requires_grad` field. Can you check `sharded_weight.requires_grad` after `shard_parameter` to see if it is True?

File: test/distributed/_sharded_tensor/ops/test_linear.py [hunk 2]
```diff
@@ -63,8 +63,38 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
...
+        sharded_loss = expected_result - torch.sum(sharded_output, dim=1)
+        sharded_loss = sharded_loss.contiguous()
+
+        # Run backward pass
+        local_loss.backward(gradient=external_grad)
+        sharded_loss.backward(gradient=external_grad)
+
+        # Verify that both weight and bias in the sharded linear has non-None grad.
+        sharded_weight = sharded_linear.weight.local_shards()[0].tensor
+        self.assertEqual(sharded_weight.requires_grad, True)
```
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/ops/test_linear.py:84]: Which of these assertions fail?

File: torch/distributed/_sharded_tensor/ops/_common.py [hunk 3]
```diff
@@ -120,15 +123,20 @@ def _result_distribute_with_col_rearrange(
...
+            *dims,
+            device=input.device,
+            dtype=combined_results.dtype,
         )
 
     # distribute the outputs using all2all.
-    dist.all_to_all_single(
-        output, combined_results, output_split_sizes=output_split_sizes, group=pg
-    )
+    output_tensor_list = all_to_all(results, group=pg, out_tensor_list=output_tensor_list)
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/_common.py:138]: Isn't this going to cause a problem since we no longer specify output_split_sizes? Also, it doesn't look like `all_to_all` has an `out_tensor_list` argument?

Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/_common.py:138]: @pritamdamania87 I added one here because we don't support that. The output split size is specified by creating the `output_tensor_list`

File: test/distributed/_sharded_tensor/ops/test_linear.py (hunk 2 above)
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_linear.py:84]: Turns out this one is true

File: test/distributed/_sharded_tensor/ops/test_linear.py (hunk 1 above)
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_linear.py:86]: yeah this is already after `shard_parameter`.

File: torch/distributed/nn/functional.py [hunk 4]
```diff
@@ -219,18 +222,24 @@ def forward(ctx, group, tensor):
...
         return (None, gx)
 
 
 class _AlltoAll(Function):
     @staticmethod
-    def forward(ctx, group, *tensors):
+    def forward(ctx, group, out_tensor_list, *tensors):
         ctx.group = group
-        out_tensor_list = [
+        ctx.input_tensor_list = [
```
Comment (fduwjj) [inline torch/distributed/nn/functional.py:234]: Storing input tensor for backward propagation helps with the error msg.

File: torch/distributed/_sharded_tensor/ops/linear.py [hunk 5]
```diff
@@ -183,27 +197,42 @@ def _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
...
+        )
 
-    gathered_input = torch.empty(input_split_sizes[rank] * world_size, input_t_size[1], device=input_t.device)
+    output_tensor_list = [
+        torch.empty(
+            input_split_sizes[rank],
+            input_t_size[1],
+            device=input.device,
+            dtype=local_shard_t.dtype,
+            requires_grad=local_shard_t.requires_grad,
```
Comment (fduwjj) [inline torch/distributed/_sharded_tensor/ops/linear.py:215]: looks like adding this does not help. Still get the None. Is there any way to print out the graph generated? cc: @pritamdamania87

**Review by fduwjj — COMMENTED** [review]
Made some changes and looks like the dimension mismatch errors are gone but still get None for grad. I will check step by step tomorrow. Not sure if I am missing anything else here.

File: torch/distributed/_sharded_tensor/ops/linear.py (hunk 5 above)
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/ops/linear.py:215]: There are some details here: https://discuss.pytorch.org/t/print-autograd-graph/692 and https://stackoverflow.com/questions/52988876/how-can-i-visualize-what-happens-during-loss-backward in terms of how you can visualize/print the graph.

File: torch/distributed/_sharded_tensor/ops/linear.py [hunk 6]
```diff
@@ -183,27 +197,41 @@ def _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
...
-        inp = torch.narrow(gathered_input, 1, r * shard_size, shard_size)
-        results.append(inp.matmul(local_shard_t))
+    for inp in output_tensor_list:
+        results.append(inp.t().contiguous().matmul(local_shard_t))
 
     # Gather all the results appropriately.
-    local_result = torch.empty_like(results[rank])
-    dist.reduce_scatter(local_result, results, group=pg)
+    output_tensor_list = [torch.empty_like(results[rank]) for _ in range(world_size)]
+    all_to_all(results, group=pg, out_tensor_list=output_tensor_list)
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/linear.py:234]: Sure.

File: torch/distributed/_sharded_tensor/ops/linear.py (hunk 5 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/linear.py:216]: Turns out that we need to call `local_shard.retain_grad()`.

File: torch/distributed/_sharded_tensor/ops/linear.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/linear.py:235]: Maybe I should send out a separate PR?

File: torch/distributed/_sharded_tensor/ops/linear.py (hunk 5 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/linear.py:216]: However, now the grad result does not match with the one from local. When print it out, the difference range from 0.2x to 2.2x.. So I doubt the correctness of the backward propagation. (But when running the original unit test -- test_all_to_all, it actually passes)

The local loss has the following graph:

> (None, 0)
> (<SumBackward1 object at 0x7f3d79327e80>, 0)
> (<AddmmBackward0 object at 0x7f3d79327ee0>, 0)
> (<AccumulateGrad object at 0x7f3d79327dc0>, 0)
> Tensor with grad found: torch.Size([11])
>  gradient: torch.Size([11])
> (None, 0)
> (<TBackward0 object at 0x7f3d79327df0>, 0)
> (<AccumulateGrad object at 0x7f3d79327eb0>, 0)
> Tensor with grad found: torch.Size([11, 16])
>  gradient: torch.Size([11, 16])

The one from sharding:

> (<MmBackward0 object at 0x7f35fb0a7880>, 0)
> (None, 0)
> (<CloneBackward0 object at 0x7f35fb0a7a30>, 0)
> (<TBackward0 object at 0x7f35fb0a77f0>, 0)
> (<CloneBackward0 object at 0x7f35fb0a7b20>, 0)
> (<SliceBackward0 object at 0x7f35fb0a7c70>, 0)
> (<AccumulateGrad object at 0x7f35fb0a7d60>, 0)
> Tensor with grad found: torch.Size([11, 16])
>  gradient: torch.Size([11, 16])
> (<AccumulateGrad object at 0x7f35fb0a7dc0>, 0)
> Tensor with grad found: torch.Size([11])
>  gradient: torch.Size([11])

This looks legit to me.. Maybe I need to look at it again tomorrow.

**Comment by fduwjj** [comment]
Update the test logic so now we are compare against the accumulated local gradients. One thing I still was not able to figure out is why the gradients from the distributed version is exactly half of that one from local. I cannot think of any logic which leads to this. And this happens for both rowwise and colwise sharding which uses different dist ops.

Create a autograd enabled `reduce_scatter`. But I encounter some issues in writing tests in test_c10d_spawn_nccl.py. Maybe reviewers can correct me where is the right place to put the test.

**Comment by fduwjj** [comment]
Add the unit test for scatter_reduce and enable the parameter of op.

**Comment by fduwjj** [comment]
Make the shareded weight detached, so that the local gradient does not get affected when doing the back propagation of the sharded one.

Now the PR is ready for review.

**Comment by fduwjj** [comment]
Add the test for the weight, so that we can make sure the code is working as expected under `max_norm` mode.

File: torch/distributed/nn/functional.py [hunk 7]
```diff
@@ -94,20 +116,23 @@ def all_gather(tensor, group=dist.group.WORLD):
     return _AllGather.apply(group, tensor)
 
 
-def all_to_all(tensors, group=dist.group.WORLD):
+def all_to_all(tensors, group=dist.group.WORLD, out_tensor_list=None):
```
Comment (pritamdamania87) [inline torch/distributed/nn/functional.py:119]: We should make the signature of this all_to_all consistent with https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L2483 and use input_split_sizes and output_split_sizes.

Ideally in nn.Linear code we should just have to replace dist.all_to_all single with the functional version and not have to change any other lines.

File: torch/distributed/_sharded_tensor/ops/linear.py [hunk 8]
```diff
@@ -183,27 +199,40 @@ def _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
...
+            dtype=local_shard_t.dtype,
+            requires_grad=local_shard_t.requires_grad,
+        )
+        for _ in range(world_size)
+    ]
 
     # Perform alltoall
-    dist.all_to_all_single(gathered_input, input_t, input_split_sizes=input_split_sizes, group=pg)
-    gathered_input = gathered_input.t()
+    all_to_all(
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/linear.py:223]: We should just have to replace `dist.all_to_all_single` with `all_to_all` with no other changes. The functional APIs should be made consistent with the core APIs in distributed_c10d.py

**Review by pritamdamania87 — COMMENTED** [review]
Great work! A few general comments:

1. Lets move the nn/functional changes to a separate PR instead of clubbing it with this PR.
2. Lets make the nn/functional APIs consistent with the ones in distributed_c10d.py. Essentially you should have to just change one line and call the functional one and not need to change the existing implementation of sharded linear at all.

File: test/distributed/_sharded_tensor/ops/test_linear.py [hunk 9]
```diff
@@ -39,15 +46,15 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
         sharded_linear = torch.nn.Linear(*linear_size)
 
         # Copy the weights and bias from local linear
-        sharded_linear.weight = local_linear.weight
-        sharded_linear.bias = local_linear.bias
+        sharded_linear.weight = torch.nn.Parameter(local_linear.weight.detach().clone())
+        sharded_linear.bias = torch.nn.Parameter(local_linear.bias)
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_linear.py:46]: > Great work! A few general comments:
> 
> 1. Lets move the nn/functional changes to a separate PR instead of clubbing it with this PR.
> 2. Lets make the nn/functional APIs consistent with the ones in distributed_c10d.py. Essentially you should have to just change one line and call the functional one and not need to change the existing implementation of sharded linear at all.

Sure, sounds good to me. I will separate them.

File: test/distributed/_sharded_tensor/ops/test_linear.py [hunk 10]
```diff
@@ -65,27 +72,62 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
...
+            local_linear.weight, sharded_dim, TEST_GPU_NUM, spec, self.rank
+        )
+        local_grad_narrowed = local_grad.narrow(sharded_dim, start_pos, chunk_size)
+        local_weight_narrowed = local_linear.weight.narrow(
+            sharded_dim, start_pos, chunk_size
+        )
+
+        # Test backward gradient calculation.
+        self.assertEqual(sharded_linear.bias.grad, local_linear.bias.grad)
+        self.assertEqual(sharded_weight, local_weight_narrowed)
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_linear.py:102]: Ahh. Originally, I wrote this to validate the narrow logic.. I agree with you for gradient, we just need to compare the grad. Let me remove it for now.

File: test/distributed/test_c10d_spawn.py [hunk 11]
```diff
@@ -85,6 +85,17 @@ def _test_allreduce_process(
...
+    @classmethod
+    def _test_reduce_scatter_process(
+            cls, rank, filename, shared_tensors, world_size, init_pg, c2p, p2c):
+        pg = init_pg(rank, filename, world_size)
+        xs = [shared_tensors[rank].clone() * (2 ** r) for r in range(world_size)]
+        y = torch.ones_like(xs[0])
+        pg.reduce_scatter(y, xs, op=c10d.ReduceOp.SUM).wait()
+        x = torch.ones_like(xs[0]) * world_size * (2 ** rank)
+        c2p.put((rank, x, y.to("cpu")))
+        p2c.get()
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/test_c10d_spawn.py:97]: Sure.

File: test/distributed/test_c10d_spawn.py [hunk 12]
```diff
@@ -85,6 +85,17 @@ def _test_allreduce_process(
         c2p.put((rank, torch.ones(2, 2) * 2, xs[0].to("cpu")))
         p2c.get()
 
+    @classmethod
+    def _test_reduce_scatter_process(
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/test_c10d_spawn.py:89]: It's being tested in the `test_c10d_nccl.py` Let me remove it for now.

File: torch/csrc/distributed/c10d/init.cpp [hunk 13]
```diff
@@ -1157,14 +1157,17 @@ that adds a prefix to each key inserted to the store.
...
                 std::vector<std::vector<at::Tensor>> inputs = {input};
-                return pg.reduce_scatter(
-                    outputs, inputs, ::c10d::ReduceScatterOptions());
+                ::c10d::ReduceScatterOptions opts;
+                opts.reduceOp = op;
+                return pg.reduce_scatter(outputs, inputs, opts);
               },
               py::arg("output_tensors"),
               py::arg("input_tensor"),
+              py::arg("op") = ::c10d::ReduceOp::SUM,
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply torch/csrc/distributed/c10d/init.cpp:1170]: Yes... It took me a while to eventually figure out this bug..

File: torch/distributed/_sharded_tensor/ops/linear.py (hunk 8 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/linear.py:223]: Sure.

File: test/distributed/test_c10d_spawn.py [hunk 14]
```diff
@@ -85,6 +85,17 @@ def _test_allreduce_process(
         c2p.put((rank, torch.ones(2, 2) * 2, xs[0].to("cpu")))
         p2c.get()
 
+    @classmethod
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/test_c10d_spawn.py:88]: Gloo does not support `scatter_reduce`. Let me see how I can test it.

File: test/distributed/test_c10d_spawn_nccl.py [hunk 15]
```diff
@@ -57,6 +57,15 @@ def test_shared_reduce_nccl(self):
                 ProcessGroupShareTensorTest._init_pg_nccl,
                 1)
 
+        @sandcastle_skip_if(not TEST_MULTIGPU, "At least 2 CUDA GPUS needed")
+        @sandcastle_skip_if(NO_NCCL, "NCCL needed")
+        def test_shared_reduce_scatter_nccl(self):
+            self._test_multiprocess(
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/test_c10d_spawn_nccl.py:63]: Looks like we don't have a unit test for nn/functional api in nccl. Let me add one here.

File: test/distributed/_sharded_tensor/ops/test_linear.py [hunk 16]
```diff
@@ -65,27 +68,62 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
...
+        sharded_loss = expected_result - torch.sum(sharded_output, dim=1)
+        self.assertEqual(local_loss, sharded_loss)
+
+        # Run backward pass
+        local_loss.backward(gradient=local_loss.detach())
+        local_grad = local_linear.weight.grad
+        sharded_loss.backward(gradient=sharded_loss.detach())
+
+        # Verify that both weight and bias in the sharded linear has non-None grad.
+        sharded_weight = sharded_linear.weight.local_shards()[0].tensor
```
Comment (wanchaol) [inline test/distributed/_sharded_tensor/ops/test_linear.py:85]: maybe we should introduce a sharded_tensor.grad attribute as we discussed, but the `.grad` probably should be a ShardedTensor as well (which contains local gradients for local shards), this can be a follow up that will make this test cleaner.

File: torch/distributed/_sharded_tensor/ops/_common.py [hunk 17]
```diff
@@ -8,6 +8,10 @@
     get_split_size,
     get_chunked_dim_size,
 )
+from torch.distributed.nn.functional import (
+    all_gather,
```
Comment (wanchaol) [inline torch/distributed/_sharded_tensor/ops/_common.py:12]: nit: linter err

File: torch/distributed/_sharded_tensor/ops/linear.py [hunk 18]
```diff
@@ -103,16 +107,24 @@ def sharded_linear(types, args, kwargs, pg):
 
     local_shard = weight.local_shards()[0].tensor
     local_shard_t = local_shard.t().contiguous()
+    local_shard.retain_grad()
```
Comment (wanchaol) [inline torch/distributed/_sharded_tensor/ops/linear.py:110]: curious why we need to have `retain_grad()` call here? aren't the linear.weight already have `.grad`?

File: test/distributed/_sharded_tensor/ops/test_linear.py (hunk 16 above)
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_linear.py:85]: sure.

File: torch/distributed/_sharded_tensor/ops/linear.py (hunk 18 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/linear.py:110]: This is something really tricky. If we don't set it, even if the `requre_grad` is true, the grad is `None`.

File: test/distributed/_sharded_tensor/ops/test_linear.py [hunk 19]
```diff
@@ -39,15 +45,15 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
         sharded_linear = torch.nn.Linear(*linear_size)
 
         # Copy the weights and bias from local linear
-        sharded_linear.weight = local_linear.weight
-        sharded_linear.bias = local_linear.bias
+        sharded_linear.weight = torch.nn.Parameter(local_linear.weight.detach().clone())
```
[the remark this answers is not in the document]
Comment (fduwjj) [reply test/distributed/_sharded_tensor/ops/test_linear.py:45]: It will raise a error complains about it. The error msg:

> TypeError: cannot assign 'torch.cuda.FloatTensor' as parameter 'weight' (torch.nn.Parameter or None expected)

**Comment by fduwjj** [comment]
Split the original PR into two and now this PR just focus on enabling Auto grad for Sharded Linear.

File: torch/distributed/_sharded_tensor/ops/linear.py [hunk 20]
```diff
@@ -183,27 +199,40 @@ def _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
...
+        )
+        for _ in range(world_size)
+    ]
 
     # Perform alltoall
-    dist.all_to_all_single(gathered_input, input_t, input_split_sizes=input_split_sizes, group=pg)
-    gathered_input = gathered_input.t()
+    all_to_all(
+        output_tensor_list,
+        list(torch.split(input_t, input_split_sizes)),
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/linear.py:225]: Seems like we have extra copies here since we are not using something like `all_to_all_single` autograd enabled function? If so, maybe we can create a gh issue for this as future improvement and add a TODO here pointing to that gh issue.

File: torch/distributed/_sharded_tensor/ops/_common.py [hunk 21]
```diff
@@ -124,14 +127,20 @@ def _result_distribute_with_col_rearrange(
...
     for idx, placement in enumerate(weight._sharding_spec.placements):
-        output_split_sizes[placement.rank()] = get_chunked_dim_size(
-            sharding_dim_size, split_size, idx
+        output_split_size = get_chunked_dim_size(sharding_dim_size, split_size, idx)
+        output_split_sizes[placement.rank()] = output_split_size
+        dims[0] = output_split_size
+        output_tensor_list[placement.rank()] = torch.empty(
+            *dims,
+            device=input.device,
+            dtype=combined_results.dtype,
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/_common.py:138]: Can't we remove `combined_results` completely and avoid the `torch.cat` above? I'm assuming you can just use `results[0].dtype` here?

File: torch/distributed/_sharded_tensor/ops/linear.py [hunk 22]
```diff
@@ -130,16 +142,20 @@ def _handle_col_wise_sharding(input, world_size, weight, local_shard_t, bias, pg
...
+            weight.size(0),
+            len(input.size()) - 1,
+            input,
+            world_size,
+            weight,
+            local_shard_t,
+            pg,
+        )
+        + bias
+    )
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/linear.py:157]: Usually its better to keep such formatting changes in a separate PR so it is easier to review.

Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/linear.py:157]: Sure. It's because the auto-format is applied to the whole doc while originally I just want to format the new code I added. I think there is a script which we can use to format all code under distributed folder so that we will not run into this kind of situation that often.

File: torch/distributed/_sharded_tensor/ops/linear.py (hunk 20 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/linear.py:225]: I will implement the `all_to_all_single` then.

File: torch/distributed/_sharded_tensor/ops/_common.py (hunk 21 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/_common.py:138]: Opps, we definitely need to remove it.

**Comment by fduwjj** [comment]
Address the comment from reviewer.

1. Change back to `all_to_all_single`.
2. Remove unnecessary code.

File: test/distributed/_sharded_tensor/ops/test_linear.py (hunk 16 above)
Comment (pritamdamania87) [reply test/distributed/_sharded_tensor/ops/test_linear.py:77]: Lets create a gh issue for ShardedTensor.grad to track this.

File: torch/distributed/_sharded_tensor/ops/_common.py [hunk 23]
```diff
@@ -124,14 +127,19 @@ def _result_distribute_with_col_rearrange(
...
+        output_tensor_list[placement.rank()] = torch.empty(
+            *dims,
+            device=input.device,
         )
 
     # distribute the outputs using all2all.
-    dist.all_to_all_single(
-        output, combined_results, output_split_sizes=output_split_sizes, group=pg
+    output = torch.cat(
+        all_to_all(output_tensor_list, results, group=pg)
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/ops/_common.py:142]: Why are we not using all_to_all_single here?

File: torch/distributed/_sharded_tensor/ops/linear.py (hunk 18 above)
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/ops/linear.py:110]: Typically this needs to be done for intermediate values: https://stackoverflow.com/questions/45988168/why-does-autograd-not-produce-gradient-for-intermediate-variables. But `local_shard` is a leaf variable and not an intermediate variable in the autograd graph.

So I'm wondering why this needs to be done. Can you try printing out the autograd graph to see why we need `retain_grad`?

**Review by pritamdamania87 — COMMENTED** [review]
Looks great, just a couple of comments.

File: torch/distributed/_sharded_tensor/ops/_common.py (hunk 23 above)
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/ops/_common.py:143]: Need to address this comment

File: torch/distributed/_sharded_tensor/ops/linear.py (hunk 18 above)
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/ops/linear.py:110]: This comment needs to be resolved too

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
@fduwjj I had a couple of more comments which need to be resolved.

File: torch/distributed/_sharded_tensor/ops/_common.py (hunk 23 above)
Comment (fduwjj) [reply torch/distributed/_sharded_tensor/ops/_common.py:143]: Sure.

**Comment by fduwjj** [comment]
Address the comment to use `all_to_all_single` for the col-wise sharding case.

The graph built for the case when there is `retain_grad` and no `retain_grad` are pretty much the same. Just one difference is that I got the following error:

> /data/sandcastle/boxes/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/_sharded_tensor/sharded_linear#binary,link-tree/torch/_tensor.py:1095: UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the .grad field to be populated for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations. (Triggered internally at  buck-out/dev/gen/caffe2/aten/generated-aten-headers-core#header-mode-symlink-tree-with-header-map,headers/ATen/core/TensorBody.h:417.)
>   return self._grad

So maybe the sharded weight is thought to be a intermediate variable?

**Comment by fduwjj** [comment]
Found the reason why we need `retain_grad ` and remove it. We need to add a context manager `no_grad` in the weight sharding `narrow` op so that the sharded weight can be the leaf.

File: torch/distributed/_sharded_tensor/__init__.py [hunk 24]
```diff
@@ -475,12 +475,17 @@ def shard_parameter(
     # Scatter the shards (use broadcast since NCCL doesn't support scatter, this is very inefficient).
     dist.broadcast(tensor, src=src_rank, group=pg)
 
-    # Reshape to get shard for this rank.
-    local_shard = tensor.narrow(
-        sharding_spec.dim,  # type: ignore[arg-type]
-        local_metadata.shard_offsets[sharding_spec.dim],  # type: ignore[union-attr, arg-type, index]
-        local_metadata.shard_sizes[sharding_spec.dim],  # type: ignore[union-attr, index]
-    ).contiguous()
+    # Narrow op is not included in the grad graph.
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/__init__.py:478]: nit: This comment might be more clear: "We don't want autograd recording here for the narrow op and 'local_shard' should be a leaf variable in the autograd graph".

**Review by pritamdamania87 — APPROVED** [review]
Awesome looks great!
